### PR TITLE
Make 6X_STABLE work with Apple M1

### PR DESCRIPTION
**Pull request by gpopt**
Make sure you have installed python 2.7:
```
brew install pyenv
pyenv install 2.7.18
```
Don't use the python 2.7 shipped with macOS.

### CODE REVIEW

File: python-dependencies.txt [hunk 1]
```diff
@@ -2,11 +2,7 @@ argparse==1.2.1
 behave==1.2.4
 epydoc==3.0.1
 logilab-astng==0.20.1
-logilab-common==0.50.1
-MarkupSafe==1.0
-mock==1.0.1
-parse==1.8.2
-psutil==4.0.0
+psutil==5.9.0
```
Comment (chrishajas) [inline python-dependencies.txt:5]: We no longer need the others?

File: gpMgmt/bin/generate-greenplum-path.sh [hunk 2]
```diff
@@ -3,14 +3,34 @@
...
+SCRIPT_DIR="$(cd "$(dirname "${SCRIPT_PATH}")" >/dev/null 2>&1 && pwd)"
 if [ ! -L "${SCRIPT_DIR}" ]; then
     GPHOME=${SCRIPT_DIR}
 else
     GPHOME=$(readlink "${SCRIPT_DIR}")
 fi
 EOF
 
+# Don't set PYTHONHOME for Apple M1
+if [ $(uname -m) == 'arm64' ]; then
```
Comment (chrishajas) [inline gpMgmt/bin/generate-greenplum-path.sh:30]: Why don't we need to set PYTHONHOME here?

File: gpMgmt/bin/generate-greenplum-path.sh [hunk 3]
```diff
@@ -3,14 +3,34 @@
 SET_PYTHONHOME="${1:-no}"
 
 cat <<"EOF"
-SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+if test -n "${ZSH_VERSION:-}"; then
```
Comment (chrishajas) [inline gpMgmt/bin/generate-greenplum-path.sh:6]: Is this related to M1?

**Review by chrishajas — COMMENTED** [review]
Thanks! We also probably want releng to review this, specifically the greenplum-path generation

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 3 above)
Comment (gpopt) [reply gpMgmt/bin/generate-greenplum-path.sh:6]: My default shell is not bash, this changes make it work on my Mac.
See https://github.com/greenplum-db/gpdb/commit/dc96f667891ef1ffcf67a534d9d7049d3bd96d29

File: python-dependencies.txt (hunk 1 above)
Comment (gpopt) [reply python-dependencies.txt:5]: Yes, we don't need them at all.
see https://github.com/greenplum-db/gpdb/commit/eb45b5e23dc92a2476b05280a8f306d1189df818

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 2 above)
Comment (gpopt) [reply gpMgmt/bin/generate-greenplum-path.sh:30]: We don't use the python at `${GPHOME}/ext/python`, we use the one installed by 
```
brew install pyenv
pyenv install 2.7.18
```
Setting PYTHONHOME will point to wrong python.

File: gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py [hunk 4]
```diff
@@ -111,11 +111,14 @@ def rm_include():
 
 libraries = ['pq']
 extra_compile_args = ['-O2']
+cpu = 'x86_64'
 
 if sys.platform == "win32":
     include_dirs.append(os.path.join(pg_include_dir_server, 'port/win32'))
 elif sys.platform == 'darwin' and sys.maxsize > 2**32:
-    extra_compile_args.extend(['-arch', 'x86_64'])
+    if os.uname()[4][:3] == 'arm':
```
Comment (fanfuxiaoran) [inline gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: why not check if it is 'arm64' like:
```
if os.uname()[4][:5] == 'arm64':
```

Comment (gpopt) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: Ah, you are right. Nice catch.

Comment (gpopt) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: BTW, what does the `os.uname()[4]` return on intel CPU? If it is `x86_64`, we can just use `os.uname()[4]` for arch.

Comment (fanfuxiaoran) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: Agree. `os.uname()[4]` is better.

Comment (fanfuxiaoran) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: `os.uname()[4] `  on intel CPU returns `x86_64`

Comment (gpopt) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]: Cool, thanks for confirming. Will use `os.uname()[4]` instead.

Comment (yihong0618) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py:119]:
```python
# better?
os.uname()[4].stratswith('arm')
```

Comment (fanfuxiaoran) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py]: how about `extra_compile_args.extend(['-arch', os.uname()[4]])`

Comment (gpopt) [reply gpMgmt/bin/pythonSrc/PyGreSQL-4.0/setup.py]: yes, did that.

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 2 above)
Comment (higuoxing) [reply gpMgmt/bin/generate-greenplum-path.sh:30]: I think the default value for `SET_PYTHONHOME` is `"no"`. See line 3 of this script.

Comment (gpopt) [reply gpMgmt/bin/generate-greenplum-path.sh:30]: In gpAux/Makefile#L605, it write yes to the value.

Comment (higuoxing) [reply gpMgmt/bin/generate-greenplum-path.sh:30]: I think that Makefile is for concourse to build gpdb. Because we are vendoring Python for gpdb in release builds, it will set the `PYTHONHOME` to `$GPHOME/ext/python`.

Comment (fanfuxiaoran) [reply gpMgmt/bin/generate-greenplum-path.sh:30]: > I think the default value for SET_PYTHONHOME is "no". See line 3 of this script.

> I think that Makefile is for concourse to build gpdb. Because we are vendoring Python for gpdb in release builds, it will set the PYTHONHOME to $GPHOME/ext/python.

Agree, seems no need to change PYTHONEHOME for mac m1.

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 3 above)
Comment (adam8157) [reply gpMgmt/bin/generate-greenplum-path.sh:6]: Didn't the shebang line work?

Comment (higuoxing) [reply gpMgmt/bin/generate-greenplum-path.sh:6]: This script is `generate-greenplum-path.sh`, so `bash` is only invoked when generating `greenplum_path.sh`.

**Comment by 24nishant** [comment]
We are able to run GPDB 6X on M1 MAC following the changes mentioned at the following link,
without these changes build was unsuccessful.

  https://github.com/gpopt/gpdb/commit/8b3b724200b09ce3742413dc7995fb100c0d13c1

Along with these changes, following updates were also made:

a) Python 2.7 was installed using following commands
    brew install pyenv
    pyenv install 2.7.18
    It was checked if python is installed at ~/.pyenv/versions/2.7.18/bin/

b) Softlink was made for python
    ln -s ~/.pyenv/versions/2.7.18/bin/python /usr/local/bin/python

c) The PATH was added in the ~/.bashrc file
    export PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Users/nishantsharma/.pyenv/versions/2.7.18/bin:/usr/local/gpdb6/bin

d) Python folder was removed from the /Library/Frameworks directory

After these changes we were able to run GPDB6 successfully.
Further, we were able to run GPDB MASTER following the standard procedure.


----------------------------------------------------------------------------------
BACKGROUND
----------------------------------------------------------------------------------

Before the above update was made, the system was showing following errors

1. On running ./README.macOS.bash command
    [~/workspace/gpdb6: {6X_STABLE} ?]$ ./README.macOS.bash
    Error
    ----
    mkdir: /usr/local/gpdb: File exists
    gdb: The x86_64 architecture is required for this software.
    Error: gdb: An unsatisfied requirement failed this build.

2. On running build6x, which inturn triggers following command:
   CFLAGS=-O0 -g3 ./configure --prefix=/usr/local/gpdb6 --config-cache --without-zstd --disable-gpcloud --enable-debug --enable-depend --enable-orca --with-python --enable-orafce --disable-gpfdist --enable-cassert
   following error were generated:

    Error
    ----
    
    CStackDescriptor.cpp:35:2: error: unknown token in expression
           GPOS_GET_FRAME_POINTER(current_frame);
           ^
    ../../../../../../src/backend/gporca/libgpos/include/gpos/utils.h:40:3: note: expanded from macro 'GPOS_GET_FRAME_POINTER'
                   GPOS_ASMFP;              \
                   ^
    ../../../../../../src/backend/gporca/libgpos/include/gpos/utils.:22:33 note: expanded from macro 'GPOS_ASMFP'
    #define GPOS_ASMFP asm volatile("movq %%rbp, %0" : "=g"(ulp));
                                   ^
    <inline asm>:1:7: note: instantiated into assembly here
           movq %rbp, [x8]
                ^
    CStackDescriptor.cpp:35:2: error: invalid operand
           GPOS_GET_FRAME_POINTER(current_frame);
           ^
    ../../../../../../src/backend/gporca/libgpos/include/gpos/utils.h:40:3: note: expanded from macro 'GPOS_GET_FRAME_POINTER'
                   GPOS_ASMFP;              \
                   ^
    ../../../../../../src/backend/gporca/libgpos/include/gpos/utils.h:22:33: note: expanded from macro 'GPOS_ASMFP'
    #define GPOS_ASMFP asm volatile("movq %%rbp, %0" : "=g"(ulp));
                                   ^
    <inline asm>:1:7: note: instantiated into assembly here
           movq %rbp, [x8]
                ^
    2 errors generated.
    make[6]: *** [CStackDescriptor.o] Error 1
    make[5]: *** [common-recursive] Error 2
    make[4]: *** [src-recursive] Error 2
    make[3]: *** [libgpos-recursive] Error 2
    make[3]: *** Waiting for unfinished jobs....
    make[2]: *** [gporca-recursive] Error 2
    make[1]: *** [all-backend-recurse] Error 2
    make: *** [all-src-recurse] Error 2


**After modifications were done:**

1.  On running ./README.macOS.bash command, the command prompt stops with following message:

    mkdir: /usr/local/gpdb: File exists
    ./README.macOS.bash: line 89: brew: command not found

2. The build is successful and GPDB6 is running

[~/workspace/gpdb6: {6X_STABLE} 6X]$ uname -a
Darwin nishants4-a01.vmware.com 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000 arm64
[~/workspace/gpdb6: {6X_STABLE} 6X]$ psql
psql (9.4.26)
Type "help" for help.

nishantsharma=# select version();
                                                                                                              version
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.4.26 (Greenplum Database 6.19.1+dev.35.ge6befcd9ba build dev) on arm-apple-darwin21.6.0, compiled by Apple clang version 14.0.0 (clang-1400.0.29.102), 64-bit compiled on Sep 21 2022 11:05:11 (with assert checking)
(1 row)

nishantsharma-# \q

[~/workspace/gpdb6: {6X_STABLE} 6X]$ gpstate -a
20220922:10:06:04:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-Starting gpstate with args: -a
20220922:10:06:04:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 6.19.1+dev.35.ge6befcd9ba build dev'
20220922:10:06:04:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-master Greenplum Version: 'PostgreSQL 9.4.26 (Greenplum Database 6.19.1+dev.35.ge6befcd9ba build dev) on arm-apple-darwin21.6.0, compiled by Apple clang version 14.0.0 (clang-1400.0.29.102), 64-bit compiled on Sep 21 2022 11:05:11 (with assert checking)'
20220922:10:06:04:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-Obtaining Segment details from master...
20220922:10:06:04:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-Gathering data from segments...
...
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-Greenplum instance status summary
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Master instance                                           = Active
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Master standby                                            = No master standby configured
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total segment instance count from metadata                = 6
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Primary Segment Status
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total primary segments                                    = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total primary segment valid (at master)                   = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total primary segment failures (at master)                = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid files missing              = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid files found                = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid PIDs missing               = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid PIDs found                 = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of /tmp lock files missing                   = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of /tmp lock files found                     = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number postmaster processes missing                 = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number postmaster processes found                   = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Mirror Segment Status
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total mirror segments                                     = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total mirror segment valid (at master)                    = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total mirror segment failures (at master)                 = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid files missing              = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid files found                = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid PIDs missing               = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of postmaster.pid PIDs found                 = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of /tmp lock files missing                   = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number of /tmp lock files found                     = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number postmaster processes missing                 = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number postmaster processes found                   = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number mirror segments acting as primary segments   = 0
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-   Total number mirror segments acting as mirror segments    = 3
20220922:10:06:07:009423 gpstate:nishants4-a01:nishantsharma-[INFO]:-----------------------------------------------------

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 2 above)
Comment (gpopt) [reply gpMgmt/bin/generate-greenplum-path.sh]: Thanks all for the comments, I had removed this change so that it won't need to check for M1.

**Review by fanfuxiaoran — APPROVED** [review]
LGTM

File: gpMgmt/bin/generate-greenplum-path.sh (hunk 3 above)
Comment (khuddlefish) [reply gpMgmt/bin/generate-greenplum-path.sh:6]: This makes me worried. I'm not sure the logistics but I might prefer to give steps for mac developers to set their shell to use Bash for developing greenplum rather than adding support for zsh

File: python-dependencies.txt (hunk 1 above)
Comment (50wu) [reply python-dependencies.txt:5]: The master changes from this [commit](https://github.com/greenplum-db/gpdb/commit/eb45b5e23dc92a2476b05280a8f306d1189df818) might not same with 6X. For example, in 6X we still use the `psutil 4.0.0`. Refer to this [Makefile](https://github.com/greenplum-db/gpdb/blob/6X_STABLE/gpMgmt/bin/Makefile)

Comment (higuoxing) [reply python-dependencies.txt:5]: Maybe we can split this change into a seperate PR, since this doesn't relate to Apple M1?
Also, logilab-common and logilab-astng both seem to be dependencies of pylint, epydoc is removed in 766ecfdc0c73b50818731e68adef4d2071ca15ed. Seems that we can remove them all.

Comment (gpopt) [reply python-dependencies.txt:5]: `psutil 4.0.0` doesn't compile on Apple M1. The gpMgmt/bin/Makefile uses `psutil 4.0.0` from `https://github.com/greenplum-db/pythonsrc-ext/tree/135d649aef8b70eb4ee8ec824f869026ac3f2b28` regardless `python-dependencies.txt`, so I guess this patch doesn't change that, thoughts?

Comment (khuddlefish) [reply python-dependencies.txt:5]: @gpopt We want to have a consistent set of dependency versions to use with gpdb. If psutil 5.9.0 works then we should bump our other dependency to be the same version. It is dangerous to use different versions in different places. Also we are in the process of merging the `pythonsrc-ext` dependencies into the gpdb repo to get rid of the submodule.

Comment (gpopt) [reply python-dependencies.txt]: OK, I have restored python-dependencies.txt to the original version.

**Comment by gpopt** [comment]
Merged to 6X_STABLE, thanks all for the review!
